### PR TITLE
MONIT-17041: Trace generation based on a full-signature trace types a…

### PR DESCRIPTION
**Pull request by dbaghdasarya**
…pproach.

### CODE REVIEW

File: src/main/java/com/wavefront/helpers/Defaults.java [hunk 1]
```diff
@@ -14,22 +15,25 @@
...
 
-  public static final List<TraceTypePattern.TagVariation> DEFAULT_MANDATORY_TAGS = List.of(
-      new TagVariation("application", List.of("Application_1")),
-      new TagVariation("cluster", List.of("us-west")),
-      new TagVariation("shard", List.of("primary")),
-      new TagVariation("service", List.of("Service_1", "Service_2")));
+  public static final List<TagVariation> DEFAULT_MANDATORY_TAGS = List.of(
+      new TagVariation("application", Set.of("Application_1")),
+      new TagVariation("cluster", Set.of("us-west")),
+      new TagVariation("shard", Set.of("primary")),
```
Comment (sirak92) [inline src/main/java/com/wavefront/helpers/Defaults.java]: Could you please define variables for string values  "application", "cluster", "shard"

File: src/main/java/com/wavefront/topology/ServiceInfo.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+import com.wavefront.datastructures.TagVariation;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Combines all service-related data. For internal use of package classes.
```
Comment (sirak92) [inline src/main/java/com/wavefront/topology/ServiceInfo.java:11]: please improve java-doc

Comment (sirak92) [reply src/main/java/com/wavefront/topology/ServiceInfo.java:11]: and also maybe it should be in another package like datastructures

File: src/main/java/com/wavefront/topology/TraceType.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import java.util.List;
+import java.util.Random;
+
+/**
+ * A full-signature traceType.
+ *
+ * @author Davit Baghdasaryan (dbagdasarya@vmware.com)
+ */
+public class TraceType {
+  private static final Random RANDOM = new Random(System.currentTimeMillis());
```
Comment (sirak92) [inline src/main/java/com/wavefront/topology/TraceType.java]: unused

File: src/main/java/com/wavefront/topology/TraceType.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
+package com.wavefront.topology;
+
+import com.wavefront.datastructures.Distribution;
+import com.wavefront.datastructures.ErrorCondition;
+
+import java.util.List;
+import java.util.Random;
+
+/**
+ * A full-signature traceType.
```
Comment (sirak92) [inline src/main/java/com/wavefront/topology/TraceType.java]: the same for ServiceInfo class

File: src/main/java/com/wavefront/topology/TraceType.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+ *
+ * @author Davit Baghdasaryan (dbagdasarya@vmware.com)
+ */
+public class TraceType {
```
Comment (sirak92) [inline src/main/java/com/wavefront/topology/TraceType.java]: also it makes sense to make fields private and use lombok for getters and setters

File: src/main/java/com/wavefront/topology/TraceTopology.java [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    List<Pair<String, String>> tags = new LinkedList<>();
+    ServiceInfo serviceInfo = serviceInfos.get(service);
+    if (serviceInfo == null) {
+      return null;
+    }
+
+    serviceInfo.tags.forEach((k, v) -> {
+      // Values in range [0..100] inclusive are possible.
+      if (RANDOM.nextInt(HUNDRED_PERCENT) + 1 <= v.percentage) {
+        tags.add(new Pair(k, v.getRandomValue()));
```
Comment (sirak92) [inline src/main/java/com/wavefront/topology/TraceTopology.java]: unchecked warning, please use diamond <> operator

File: src/main/java/com/wavefront/topology/TraceTopologySanitizer.java [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+import java.util.logging.Logger;
+
+import static com.wavefront.helpers.Defaults.DEFAULT_MANDATORY_TAGS;
+import static com.wavefront.helpers.Defaults.SERVICE;
+
+/**
+ * Sanitizer class which converts incorrect input values to default ones.
+ *
+ * @author Davit Baghdasaryan (dbaghdasarya@vmware.com)
+ */
```
Comment (sirak92) [inline src/main/java/com/wavefront/topology/TraceTopologySanitizer.java:20]: Maybe makes sense to move to helpers package

File: src/main/java/com/wavefront/topology/ServiceInfo.java (hunk 2 above)
Comment (dbaghdasarya) [reply src/main/java/com/wavefront/topology/ServiceInfo.java:11]: Java-doc improved.
Regarding moving: I put TraceTopology, TraceTopologySanitizer, ServiceInfo in one package for preserving encapsulation principles. These three classes have wide access to each other, but I don't willing to provide the same access level to others.

File: src/main/java/com/wavefront/topology/TraceType.java (hunk 3 above)
Comment (dbaghdasarya) [reply src/main/java/com/wavefront/topology/TraceType.java]: Deleted

File: src/main/java/com/wavefront/topology/TraceType.java (hunk 5 above)
Comment (dbaghdasarya) [reply src/main/java/com/wavefront/topology/TraceType.java]: I will keep in mind this for future work, but this is a stupid data structure without any functionality. If you don't mind I prefer to leave it with public modifiers.

File: src/main/java/com/wavefront/topology/TraceType.java (hunk 4 above)
Comment (dbaghdasarya) [reply src/main/java/com/wavefront/topology/TraceType.java]: Moved

File: src/main/java/com/wavefront/topology/TraceTopology.java (hunk 6 above)
Comment (dbaghdasarya) [reply src/main/java/com/wavefront/topology/TraceTopology.java]: Fixed

**Pull request merged**